ghcide/test/exe/InlayHintTests.hs

```diff
@@ -0,0 +1,179 @@
+{-# LANGUAGE ExplicitNamespaces #-}
+
+module InlayHintTests (tests) where
+
+import           Config                      (mkIdeTestFs, testWithDummyPlugin,
+                                              testWithDummyPluginEmpty)
+import           Control.Monad               (void)
+import           Control.Monad.IO.Class      (MonadIO (liftIO))
+import qualified Data.Aeson                  as A
+import           Data.Maybe                  (mapMaybe)
+import qualified Data.Text                   as T
+import           Language.LSP.Protocol.Types (InlayHint (..),
+                                              Position (Position),
+                                              Range (Range, _end, _start),
+                                              TextDocumentIdentifier (TextDocumentIdentifier),
+                                              TextEdit (TextEdit, _newText, _range),
+                                              UInt,
+                                              VersionedTextDocumentIdentifier (_uri),
+                                              type (|?) (..))
+import           Language.LSP.Test           (applyEdit, createDoc,
+                                              documentContents, getInlayHints,
+                                              openDoc, setConfigSection)
+import           Test.Hls                    (Assertion, Session, expectFail,
+                                              waitForTypecheck)
+import           Test.Hls.FileSystem         (copyDir)
+import           Test.Tasty                  (TestTree, testGroup)
+import           Test.Tasty.HUnit            ((@=?), (@?=))
+
+tests :: TestTree
+tests = testGroup "inlay hints"
+  [ whereInlayHintsTests
+  ]
+
+whereInlayHintsTests :: TestTree
+whereInlayHintsTests = testGroup "add signature for where clauses"
+  [ testWithDummyPluginEmpty "No where inlay hints if disabled" $ do
+      let content = T.unlines
+            [ "module Sigs where"
+            , "f :: b"
+            , "f = undefined"
+            , "  where"
+            , "    g = True"
+            ]
+          range = Range { _start = Position 4 0
+                        , _end = Position 4 1000
+                        }
+      doc <- createDoc "Sigs.hs" "haskell" content
+      setConfigSection "haskell" (createConfig False)
+      inlayHints <- getInlayHints doc range
+      liftIO $ length inlayHints @?= 0
+  , testGroup "apply EditText"
+    [ editTest "Simple"
+    , editTest "Tuple"
+    , editTest "Inline"
+    , editTest "Infix"
+    , editTest "Operator"
+    , expectFail $ editTest "ScopedTypeVariables"
+    , editTest "Nest"
+    , editTest "NoLens"
+    , expectFail $ editTest "Typeclass"
+    , editTest "Qualified"
+    ]
+  , testGroup "apply EditText"
+    [ hintTest "Simple" $ (@=?)
+      [defInlayHint { _position = Position 5 9
+                    , _label = InL ":: Bool"
```

Comment:
I wonder if we should display the hints on the previous line rather than right after the binding. i.e. make the hint look like the edit currently does.

Advantages:
1. It makes the effect of applying the hint align with how the hint looks, which is I think what we're supposed to do.
2. It makes the situation less bad if the type is very long. That might already be annoying, but it could be even more annoying if it's right in the middle of something you're trying to read. Maybe we should look at what other language servers do?

Reply:
I don't think inlay hints can do this.
I found this issue: https://github.com/microsoft/language-server-protocol/issues/1821.

AFAIK, standard inlay hints cannot be in virtual line.
It's more like a fixed position code lens?